File: adc09.ml
#require "brulib"

open Brulib

(* let file = "test09.txt" *)
let file = "input09.txt"

let follow_once (tx, ty) (hx, hy) =
  let dx = tx - hx and dy = ty - hy in
  if abs dx = 2 || abs dy = 2 then
    (hx + (dx / 2), hy + (dy / 2))
  else (hx + dx, hy + dy)

let dir = function
  | 'L' -> (-1, 0)
  | 'R' -> (1, 0)
  | 'U' -> (0, 1)
  | 'D' -> (0, -1)
  | _ -> failwith "Wrong direction"

let follow t = Gen.genfold follow_once (0, 0) t

let data () =
  Gen.scan_file file "%c %d" (fun c d -> (dir c, d))

let head () =
  Gen.of_fun (fun yield ->
      let x = ref 0 and y = ref 0 in
      yield (!x, !y);
      Gen.iter
        (fun ((dx, dy), amp) ->
          for _ = 1 to amp do
            x := !x + dx;
            y := !y + dy;
            yield (!x, !y)
          done)
        (data ()))

let count t =
  let h = Hashtbl.create 100 in
  Gen.iter (fun v -> Hashtbl.replace h v ()) t;
  Hashtbl.length h

let _ =
  Printf.printf "Part One: %d\n"
    (Gen.(head () |> follow) |> count)

let _ =
  let pos = ref (head ()) in
  for _ = 1 to 9 do
    pos := follow !pos
  done;
  Printf.printf "Part Two: %d\n" (count !pos)

File: adc09.py
import math

# file = "test09.txt"
# file = "test09_2.txt"
file = "input09.txt"

dirs = {"L": (0, -1), "R": (0, 1), "U": (1, 1), "D": (1, -1)}


# I use copysign since -1 // 2 == -1

def follow(head, tail):
    delta = [tail[i] - head[i] for i in range(2)]
    if max(abs(delta[0]), abs(delta[1])) == 2:
        return tuple(head[i] + math.copysign(abs(delta[i]) // 2, delta[i])
                     for i in range(2))
    return tuple(head[i] + delta[i] for i in range(2))


def check(a, b):
    for i in range(2):
        assert abs(a[i] - b[i]) <= 1


with open(file, "r") as f:
    head = [0, 0]
    traj = [tuple(head)]
    for line in f:
        dir = line[0]
        d, f = dirs[dir]
        amp = int(line[2:])
        for _ in range(amp):
            head[d] += f
            traj.append(tuple(head))
    for i in range(9):
        tail = (0, 0)
        traj2 = [tail]
        for h in traj:
            old = tail
            tail = follow(h, tail)
            traj2.append(tail)
            check(h, tail)
            check(old, tail)
        traj = traj2
    print(len(set(traj)))
